### PR TITLE
Remove remaining hard-coded profile fields

### PR DESCRIPTION
**Pull request by jacobherrington**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/policies/user_policy.rb [hunk 1]
```diff
@@ -20,7 +20,8 @@ class UserPolicy < ApplicationPolicy
     email_follower_notifications
     email_membership_newsletter
     email_mention_notifications
-    email_newsletter email_public
+    email_newsletter
+    email_public
```
Comment (citizen428) [inline app/policies/user_policy.rb:24]: This is now on `Users::Setting` and called `display_email_on_profile`. It's probably best to wait for Arit's PR to be merged then rebase this against `main`.

**Review by citizen428 — APPROVED** [review]

**Comment by citizen428** [comment]
Oops, I had accidentally approved this previously, so re-requested my own review now.

File: app/views/articles/_user_metadata.html.erb [hunk 2]
```diff
@@ -43,43 +43,47 @@
...
-          </div>
-          <div class="value">
-            <%= @user.employment_title %>
-            <% if @user.employer_name.present? %>
-              <span class="opacity-50"> at </span>
-              <% if @user.employer_url.present? %>
-                <a href="<%= @user.employer_url %>" target="_blank" rel="noopener"><%= @user.employer_name %></a>
-              <% else %>
-                <%= @user.employer_name %>
+      <% if ForemInstance.dev_to? %>
```
Comment (citizen428) [inline app/views/articles/_user_metadata.html.erb:46]: @benhalpern Should we slowly phase out these special cases?

We have the profile fields and their information, we can still render them just like we render any locally defined profile fields (i.e. on the profile page) but maybe we get rid of special treatment for DEV for the right sidebar?

File: app/views/articles/_user_metadata.html.erb [hunk 3]
```diff
@@ -43,43 +43,47 @@
...
-      <% if @user.education.present? %>
-        <li>
-          <div class="key">
-            Education
-          </div>
-          <ul class="value">
-            <%= @user.education %>
-          </ul>
-        </li>
+      <% if ForemInstance.dev_to? %>
```
Comment (citizen428) [inline app/views/articles/_user_metadata.html.erb:76]: @benhalpern Same question as above.

File: app/views/articles/_user_metadata.html.erb (hunk 2 above)
Comment (benhalpern) [reply app/views/articles/_user_metadata.html.erb:46]: I much prefer `employment_title` being special vs DEV being special. So definitely in favor of removing `ForemInstance.dev_to?` as a solution of any kind.

Employment title is somewhat universal and may be present in any professional network.

_However_, if that leaves a bad taste in your mouth from a complexity perspective..... We could ditch all these fields and run a migration where they logically fall into one single "Work" field for display.

So this:

Screen Shot 2021-06-17 at 6 35 11 PM

Becomes this:

Screen Shot 2021-06-17 at 6 36 16 PM

Basically we migrate to a single work field which takes an arbitrary string. That's pretty okay to me. We don't lose a lot if we do that right.

File: app/views/articles/_user_metadata.html.erb (hunk 3 above)
Comment (benhalpern) [reply app/views/articles/_user_metadata.html.erb:76]: Yeah, definitely don't need to rely on anything special about DEV here in this case.

File: app/views/articles/_user_metadata.html.erb (hunk 2 above)
Comment (citizen428) [reply app/views/articles/_user_metadata.html.erb:46]: @benhalpern Thanks for the feedback so far! Just to clarify: would the new `work` field be 

1. A `ProfileField` on DEV only (which means we have to create it and it gets stores as JSONB in the `data` column of the profiles) or
2. available on all Forems in which case we'd add it as a column to the `profiles` table.

Comment (benhalpern) [reply app/views/articles/_user_metadata.html.erb:46]: @citizen428 This can be a custom profile field, so for compatibility purposes it should be migrated to for _existing_ Forems (not just DEV), as a matter of maintaining backwards compatibility for those who have been using `employer_title,name,url`... So we should migrate data to this column not only on DEV, but we don't need to make "work" a base profile field like `location`. 

Here is a screenshot from the.community.club where folks use the employment trifecta at the moment)...

Screen Shot 2021-06-19 at 5 36 12 PM

Comment (citizen428) [reply app/views/articles/_user_metadata.html.erb:46]: @benhalpern Agreed, a single "Work" field with migrated data makes sense. Just so we are 100% on the same page, we essentially have two options here:

1. We add this as a column to `profiles`, same as `location`, `summary` and `website_url`. This would guarantee that the field always exists so we can use it in more contexts but admins can't disable it.
2. We add it as a `ProfileField`. This would give admins the option to remove it which means we could only use it in very limited contexts, like the profile page itself.

Please let me know what your preference is here.

File: app/views/articles/_user_metadata.html.erb (hunk 3 above)
Comment (citizen428) [reply app/views/articles/_user_metadata.html.erb:76]: @benhalpern Again, trying to ensure we fully are in agreement about the ramifications: if we don't want to special case this for DEV we would be completely removing it as there's no guarantee that Forem instances have a profile field like this or that it has this exact name.

File: app/views/articles/_user_metadata.html.erb (hunk 2 above)
Comment (benhalpern) [reply app/views/articles/_user_metadata.html.erb]: @citizen428 Let's go with **Option 2**. We'll migrate to a `ProfileField` which is for display on the profile, and no guarantee that it is present on any Forem.

It is for profile display-purposes only and/or any additional functionality that is generally applied to `ProfileField` in the future.

File: app/views/articles/_user_metadata.html.erb (hunk 3 above)
Comment (benhalpern) [reply app/views/articles/_user_metadata.html.erb]: @citizen428 — Yes, `education` can be entirely optional and functionally equivalent to `astrology_sign` for a Zodiac Forem or `favorite_team` for an NBA Forem.

***

(As a total aside, we may want to make room for a "select input" type of profile field because my above examples would both call for a finite number of options— There could definitely be a lot of value in profile fields which act as structured data instead of free text.)

Comment (citizen428) [reply app/views/articles/_user_metadata.html.erb]: @benhalpern Yes, we could add that. I may actually open up a stage 0 RFC for this though as we may want to think about how to store the options.
